File: bindings/csharp/FastParse/FastParseField.cs
namespace FastParse;

/// <summary>
/// Field flags that control which AST fields FastParse includes in the output.
/// </summary>
[Flags]
public enum FastParseField : uint
{
    /// <summary>Use the native default field set.</summary>
    DefaultAll = 0,

    /// <summary>Include the node id.</summary>
    Id = 1u << 0,

    /// <summary>Include the parent node id.</summary>
    ParentId = 1u << 1,

    /// <summary>Include the Tree-sitter grammar rule name.</summary>
    Rule = 1u << 2,

    /// <summary>Include the original source bytes/text for the node.</summary>
    Text = 1u << 3,

    /// <summary>Include start/end line and column positions.</summary>
    Range = 1u << 4,

    /// <summary>Include start/end byte offsets.</summary>
    ByteRange = 1u << 5,

    /// <summary>Include the number of child nodes.</summary>
    ChildCount = 1u << 6,

    /// <summary>Include child rule/text summaries when supported by the output format.</summary>
    Children = 1u << 7,

    /// <summary>Include Tree-sitter parse diagnostic fields such as isError, isMissing, and hasError.</summary>
    Diagnostics = 1u << 8,

    /// <summary>Include the Tree-sitter query capture name, for example method.name.</summary>
    CaptureName = 1u << 9,

    /// <summary>Include the matched Tree-sitter query pattern index.</summary>
    PatternIndex = 1u << 10,

    /// <summary>Include the Tree-sitter field name connecting the node to its direct parent.</summary>
    FieldName = 1u << 11,

    /// <summary>Include the zero-based Tree-sitter child index within the direct parent.</summary>
    ChildIndex = 1u << 12,

    /// <summary>Include whether the node is named by the Tree-sitter grammar.</summary>
    Named = 1u << 13,

    /// <summary>Include the zero-based node depth, where the root is depth zero.</summary>
    Depth = 1u << 14,

    /// <summary>Request all supported fields.</summary>
    All = 0xFFFFFFFFu
}

File: bindings/java/fastparse/src/main/java/dev/fastparse/FastParseField.java
package dev.fastparse;

/** Bit flags controlling which fields FastParse includes in AST/query output. */
public enum FastParseField {
    ID(1L << 0),
    PARENT_ID(1L << 1),
    RULE(1L << 2),
    TEXT(1L << 3),
    RANGE(1L << 4),
    BYTE_RANGE(1L << 5),
    CHILD_COUNT(1L << 6),
    CHILDREN(1L << 7),
    DIAGNOSTICS(1L << 8),
    CAPTURE_NAME(1L << 9),
    PATTERN_INDEX(1L << 10),
    FIELD_NAME(1L << 11),
    CHILD_INDEX(1L << 12),
    NAMED(1L << 13),
    DEPTH(1L << 14),
    ALL(0xFFFFFFFFL);

    private final long mask;

    FastParseField(long mask) {
        this.mask = mask;
    }

    public long mask() {
        return mask;
    }

    public static long maskOf(FastParseField... fields) {
        if (fields == null || fields.length == 0) {
            return 0L;
        }
        long value = 0L;
        for (FastParseField field : fields) {
            if (field != null) {
                value |= field.mask;
            }
        }
        return value;
    }
}
